bundles/org.openhab.binding.comfoair/src/main/java/org/openhab/binding/comfoair/internal/ComfoAirCommandType.java

```diff
@@ -0,0 +1,862 @@
+        this.key = key;
+        this.data_type = data_type;
+        this.read_command = read_command;
+        this.read_reply_command = read_reply_command;
+        this.read_reply_data_pos = read_reply_data_pos;
+    }
+
+    public static class Constants {
+        public static final int REQUEST_GET_INPUTS = 0x03;
+        public static final int REPLY_GET_INPUTS = 0x04;
+        public static final int REQUEST_GET_FAN = 0x0b;
+        public static final int REPLY_GET_FAN = 0x0c;
+        public static final int REQUEST_GET_ANALOG_VOLTS = 0x13;
+        public static final int REPLY_GET_ANALOG_VOLTS = 0x14;
+        public static final int REQUEST_GET_FIRMWARE = 0x69;
+        public static final int REPLY_GET_FIRMWARE = 0x6a;
+        public static final int REQUEST_GET_SENSORS = 0x97;
+        public static final int REPLY_GET_SENSORS = 0x98;
+        public static final int REQUEST_SET_LEVEL = 0x99;
+        public static final int REQUEST_SET_RS232 = 0x9b;
+        public static final int REPLY_SET_RS232 = 0x9c;
+        public static final int REQUEST_GET_ANALOGS = 0x9d;
+        public static final int REPLY_GET_ANALOGS = 0x9e;
+        public static final int REQUEST_SET_ANALOGS = 0x9f;
+        public static final int REQUEST_GET_DELAYS = 0xc9;
+        public static final int REPLY_GET_DELAYS = 0xca;
+        public static final int REQUEST_SET_DELAYS = 0xcb;
+        public static final int REQUEST_GET_FAN_LEVEL = 0xcd;
+        public static final int REPLY_GET_FAN_LEVEL = 0xce;
+        public static final int REQUEST_SET_FAN_LEVEL = 0xcf;
+        public static final int REQUEST_GET_TEMPS = 0xd1;
+        public static final int REPLY_GET_TEMPS = 0xd2;
+        public static final int REQUEST_SET_TEMPS = 0xd3;
+        public static final int REQUEST_GET_STATES = 0xd5;
+        public static final int REPLY_GET_STATES = 0xd6;
+        public static final int REQUEST_SET_STATES = 0xd7;
+        public static final int REQUEST_GET_ERRORS = 0xd9;
+        public static final int REPLY_GET_ERRORS = 0xda;
+        public static final int REQUEST_SET_RESETS = 0xdb;
+        public static final int REQUEST_GET_HOURS = 0xdd;
+        public static final int REPLY_GET_HOURS = 0xde;
+        public static final int REQUEST_GET_BYPASS = 0xdf;
+        public static final int REPLY_GET_BYPASS = 0xe0;
+        public static final int REQUEST_GET_PREHEATER = 0xe1;
+        public static final int REPLY_GET_PREHEATER = 0xe2;
+        public static final int REQUEST_GET_RF = 0xe5;
+        public static final int REPLY_GET_RF = 0xe6;
+        public static final int REQUEST_GET_EWT = 0xeb;
+        public static final int REPLY_GET_EWT = 0xec;
+        public static final int REQUEST_SET_EWT = 0xed;
+    }
+
+    /**
+     * @return command key
+     */
+    public String getKey() {
+        return key;
+    }
+
+    /**
+     * @return data type for this command key
+     */
+    public @Nullable ComfoAirDataType getDataType() {
```

Comment:
Instead of having to resort to reflection in order to create instances, you could instead create static singlton instances of all your datatype classes so you can just return those singleton instances here. There is no reason you can't reuse your datatype instances since they are stateless.

Although it might be the case that the compiler won't let you use singleton instances in the enum constructor, in that case you can pass a lambda for the constructor instead like `DataTypeBoolean::new`, the enum constructor would then use `Supplier<? extends ComfoAirDataType>` as a parameter.

Also, I'm pretty sure this method should never return null normally, so remove the `@Nullable`. Instead of returning null you should throw some kind of runtime exception, not that you would need to if you use singletons.

```suggestion
    public ComfoAirDataType getDataType() {
```

Reply:
Never worked with Singletons before. Could you give an example how this would look like?

Reply:
I think I figured out how to use Singletons for the datatype classes. However, since I can't have a `static getInstance()` method on the `ComfoAirDataType` interface, I cannot figure out how to call the method here.
Any advice?

Reply:
You can initialize your ComfoAirCommandType enum types with the singleton instances.
So 
```
public enum ComfoAirCommandType{

...
EXAMPLE(..., ExampleComfoAirDataType.getInstance(), ...);
...
private final ComfoAirDataType dataType;
...

private ComfoAirCommandType(..., ComfoAirDataType dataType, ...){
    ...
    this.dataType = dataType;
    ...
}
...
public ComfoAirDataType getDataType() {
     return dataType;
}
...
} 
```